File: 332. Reconstruct Itinerary/C++/Solution.cpp
class Solution {
public:
    vector<string> findItinerary(vector<pair<string, string>> tickets) {
        class Utils {
        public:
            bool dfs(unordered_map<string, vector<string>> &travels, vector<string> &result, string start, int length) {
                result.push_back(start);
                if (travels.find(start) == travels.end() || travels[start].empty()) {
                    if (result.size() == length) {
                        return true;
                    }
                    result.pop_back();
                    return false;
                }
                for (int i = 0; i < travels[start].size(); i++) {
                    string newStart = travels[start][i];
                    travels[start].erase(travels[start].begin() + i);
                    if (dfs(travels, result, newStart, length)) {
                        return true;
                    }
                    travels[start].insert(travels[start].begin() + i, newStart);
                }
                result.pop_back();
                return false;
            }
        };
        unordered_map<string, vector<string>> travels;
        for (auto &ticket : tickets) {
            travels[ticket.first].push_back(ticket.second);
        }
        for (auto &item : travels) {
            sort(item.second.begin(), item.second.end());
        }
        string start = "JFK";
        vector<string> result;
        Utils utils;
        utils.dfs(travels, result, start, tickets.size() + 1);
        return result;
    }
};

File: 332. Reconstruct Itinerary/Python/Solution.py
class Solution(object):
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        def dfs(travels, result, start, length):
            result.append(start)
            if start not in travels or not travels[start]:
                if len(result) == length:
                    return True
                result.pop()
                return False
            for i in range(len(travels[start])):
                new_start = travels[start].pop(i)
                if dfs(travels, result, new_start, length):
                    return True
                travels[start].insert(i, new_start)
            result.pop()

        travels = {}
        for ticket in tickets:
            if ticket[0] not in travels:
                travels[ticket[0]] = []
            travels[ticket[0]].append(ticket[1])
        for item in travels.values():
            item.sort()
        start = 'JFK'
        result = []
        dfs(travels, result, start, len(tickets) + 1)
        return result
